### ISSUE TITLE
inginious-webapp command fails after configuration

### ISSUE THREAD

**Issue by mbifulco** [UCL-INGI/INGInious#257]
After running configuration of inginious-webapp locally on my mac following [these](http://inginious.readthedocs.io/en/v0.4/install_doc/installation.html) instructions (and circumventing a bug around downloading docker images that I documented in #256), the `inginious-webapp` command fails with a `pymongo.errors.ServerSelectionTimeoutError`:

```
Traceback (most recent call last):
  File "/usr/local/bin/inginious-webapp", line 59, in <module>
    application, close_app_func = inginious.frontend.webapp.app.get_app(config)
  File "/usr/local/lib/python3.6/site-packages/inginious/frontend/webapp/app.py", line 125, in get_app
    user_manager = UserManager(web.session.Session(appli, MongoStore(database, 'sessions')), database, config.get('superadmins', []))
  File "/usr/local/lib/python3.6/site-packages/inginious/frontend/common/session_mongodb.py", line 87, in __init__
    self.collection.ensure_index(_atime)
  File "/usr/local/lib/python3.6/site-packages/pymongo/collection.py", line 1604, in ensure_index
    self.__create_index(keys, kwargs)
  File "/usr/local/lib/python3.6/site-packages/pymongo/collection.py", line 1459, in __create_index
    with self._socket_for_writes() as sock_info:
  File "/usr/local/Cellar/python3/3.6.3/Frameworks/Python.framework/Versions/3.6/lib/python3.6/contextlib.py", line 81, in __enter__
    return next(self.gen)
  File "/usr/local/lib/python3.6/site-packages/pymongo/mongo_client.py", line 868, in _get_socket
    server = self._get_topology().select_server(selector)
  File "/usr/local/lib/python3.6/site-packages/pymongo/topology.py", line 214, in select_server
    address))
  File "/usr/local/lib/python3.6/site-packages/pymongo/topology.py", line 189, in select_servers
    self._error_message(selector))
pymongo.errors.ServerSelectionTimeoutError: localhost:27017: [Errno 61] Connection refused
```

I've configured inginious using the `inginious-install webapp` command, using mostly default selections:

```
$ inginious-install webapp

--- BACKEND CONFIGURATION ---
Do you have a local docker daemon (on Linux), do you use docker-machine via a local machine, or do you use Docker for macOS? [yes]:
(INFO) If you use docker-machine on macOS, please see http://inginious.readthedocs.io/en/latest/install_doc/troubleshooting.html
(INFO) Backend chosen: local. Testing the configuration.
Maximum concurrency (number of tasks running simultaneously). Leave it empty to use the number of CPU of your host. []:
What is the external hostname/address of your machine? You can leave this empty and let INGInious autodetect it. []:
(INFO) You can now enter the port range for the remote debugging feature of INGInious. Please verify that these ports are open in your firewall. You can leave this parameters empty, the default is 64100-64200
Beginning of the range []:
(INFO) - Asking Docker some info
(INFO) - Successfully got info from Docker. Docker connection works.

--- MONGODB CONFIGURATION ---
(INFO) Trying default configuration
Successfully connected to MongoDB. Do you want to edit the configuration anyway? [no]:

--- TASK DIRECTORY ---
Please choose a directory in which to store the course/task files. By default, the tool will put them in the current directory
Task directory [.]:
Demonstration tasks can be downloaded to let you discover INGInious.
Would you like to download them ? [yes]: yes
(INFO) Successfully downloaded and copied demonstration tasks.

--- CONTAINERS ---
The tool will now propose to download some base container image for multiple languages.
Please note that the download of these images can take a lot of time, so choose only the images you need
Download default (Default container. For Bash and Python 2 tasks) ? [yes]:
Download cpp (Contains gcc and g++ for compiling C++) ? [no]:
Download java7 (Contains Java 7) ? [no]:
Download java8scala (Contains Java 8 and Scala) ? [no]:
Download mono (Contains Mono, which allows to run C#, F# and many other languages) ? [no]:
Download oz (Contains Mozart 2, an implementation of the Oz multi-paradigm language, made for education) ? [no]:
Download php (Contains PHP 5) ? [no]:
Download pythia0compat (Compatibility container for Pythia 0) ? [no]:
Download pythia1compat (Compatibility container for Pythia 1) ? [no]:
Download r (Can run R scripts) ? [no]:
Download sekexe (Can run an user-mode-linux for advanced tasks) ? [no]:
(INFO) Connecting to the local Docker daemon...
(INFO) Downloading image ingi/inginious-c-default. This can take some time.
(ERROR) An error occurred while pulling the image: 'DockerClient' object has no attribute 'pull' In Docker SDK for Python 2.0, this method is now on the object APIClient. See the low-level API section of the documentation for more details..
Do you want to manually add some images? [no]:
(INFO) Configuration of the containers done.

--- MISC ---
Use minified javascript? (Useful in production, but should be disabled in dev environment) [yes]: no

--- BACKUP DIRECTORY ---
Please choose a directory in which to store the backup files. By default, the tool will them in the current directory
Backup directory [.]: backup

--- BATCH CONTAINERS ---
The web app supports a feature called 'batch containers' that allows to apply batch operations on all submissions of a course.
Please note that the download of these images can take a lot of time, so choose only the images you need
Download ingi/inginious-b-jplag (Plagiarism detection) ? [no]: no
(INFO) Connecting to the local Docker daemon...
To you want to manually add some images? [no]: no
(INFO) Configuration of the batch containers done.

--- AUTHENTIFICATION ---
(INFO) You can choose an authentication plugin between:
(INFO) - 1. Test auth plugin. This plugin allows you to test locally INGInious, using password defined in the config file.
(INFO) - 2. DB auth plugin. This plugin stores users on the web app database and supportsself registration
(INFO) - 3. LDAP auth plugin. This plugin allows to connect to a distant LDAP host.
Enter the corresponding number to your choice [1]: 1
Authentication method name (will be displayed on the login page) [Demo]:
Let's add some users
Username [test]:
Password [test]:
Would you like to add another user? [no]: no
Would you like to add another auth method? [no]:
(INFO) You may want to add additional plugins to the configuration file.
(INFO) We will now add superadmin users.
Enter the login of a superadmin (leave empty to skip this step) []:

--- REMOTE DEBUGGING - IN BROWSER ---
(INFO) If you want to activate the remote debugging of task in the users' browser, you have to install separately INGInious-xterm, which is available on Github, according to the parameters you have given for the hostname and the port range given in the configuration of the remote debugging.
(INFO) You can leave the following question empty to disable this feature; remote debugging will still be available, but not in the browser.
Please indicate the link to your installation of INGInious-xterm (for example: https://your-hostname.com:8080). []:

--- END ---
(INFO) Successfully written the configuration file
```

Is this something you've seen before? What might I do to try to get around this?

Thank you!

**Comment by anthonygego** [comment]
Are you running the ``inginious-webapp`` script using the same user as for the ``inginious-install`` one ? ``Connection refused`` typically means either that nothing is listening on the port (the mongo service is down) either that a firewall is blocking the port. If you can use the install script, the mongo service should be running fine. Therefore, I guess this is a permission issue.

**Comment by mbifulco** [comment]
Yes, it's being run by the same user.  This also fails if I run with `sudo`.  I've also confirmed that there's nothing blocking the running port.  The `inginious-install` command was able to detect mongo, so I believe it is connecting to the db just fine as well.

**Comment by anthonygego** [comment]
It seems that for an unknown reason the pymongo api does not perform (anymore ?) the connection even when the ``connect=True`` argument is passed to the ``MongoClient`` constructor. Therefore, the connection is performed at the time of the first operation, which is, in this case, the ``sessions`` collection creation. So the user is actually unable to connect to the Mongo server. Are you able to run the Mongo shell without any issue ?

**Comment by mbifulco** [comment]
Ah, ok.  So this is definitely my own fault, but I think there's an opportunity for your docs to be a bit clearer.  After installing homebrew with mongo, I checked that it could be run with `mongod` (which took a few extra steps that I found in mongo's documentation).  Once I saw that mongod ran correctly, I closed that process... I guess I had assumed that inginious would make sure it was running if needed. 

Perhaps add a bit about "make sure mongo is running in a separate process/terminal window" here?

> OS X 10.9+
> 
> We use brew to install some packages. Packages are certainly available too via macPorts.
> 
> $ brew install mongodb
> $ brew install python3
> Follow the instruction of brew to enable mongodb.

**Comment by mbifulco** [comment]
So - I'm closer now.  I've got the inginious-webapp process running and communicating with mongo,
```bash
$ inginious-webapp
2017-10-12 10:35:23,181 - inginious.webapp - INFO - http://localhost:8080/
2017-10-12 10:35:23,181 - inginious.agent.docker - INFO - Agent started
2017-10-12 10:35:23,182 - inginious.agent.docker - INFO - Saying hello to the backend
2017-10-12 10:35:23,183 - inginious.agent.mcq - INFO - Agent started
2017-10-12 10:35:23,183 - inginious.agent.mcq - INFO - Saying hello to the backend
2017-10-12 10:35:23,183 - inginious.backend - INFO - Backend started
2017-10-12 10:35:23,192 - inginious.client - INFO - Connecting to backend
2017-10-12 10:35:23,194 - inginious.backend - INFO - Agent b'\x00\x1c\x06\xda\xc8' (Docker - Local agent) said hello
2017-10-12 10:35:23,194 - inginious.backend - INFO - Agent b'\x00\x1c\x06\xda\xc9' (MCQ - Local agent) said hello
2017-10-12 10:35:23,194 - inginious.backend - INFO - New client connected b'\x00\xd6\xe5\t\xfe'
2017-10-12 10:35:23,195 - inginious.client - INFO - Updated containers and batch containers
```

However, when I try to access it from my browser (`localhost:8080`) I get an error about `Cookie`:

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/web/wsgiserver/wsgiserver3.py", line 1089, in communicate
    req.respond()
  File "/usr/local/lib/python3.6/site-packages/web/wsgiserver/wsgiserver3.py", line 877, in respond
    self.server.gateway(self).respond()
  File "/usr/local/lib/python3.6/site-packages/web/wsgiserver/wsgiserver3.py", line 1980, in respond
    response = self.req.server.wsgi_app(self.env, self.start_response)
  File "/usr/local/lib/python3.6/site-packages/inginious/common/log.py", line 47, in __call__
    return self.app(environ, xstart_response)
  File "/usr/local/lib/python3.6/site-packages/inginious/frontend/common/static_middleware.py", line 64, in __call__
    return self.app(environ, start_response)
  File "/usr/local/lib/python3.6/site-packages/web/application.py", line 295, in wsgi
    result = self.handle_with_processors()
  File "/usr/local/lib/python3.6/site-packages/web/application.py", line 267, in handle_with_processors
    return process(self.processors)
  File "/usr/local/lib/python3.6/site-packages/web/application.py", line 264, in process
    raise self.internalerror()
  File "/usr/local/lib/python3.6/site-packages/web/application.py", line 545, in internalerror
    return debugerror()
  File "/usr/local/lib/python3.6/site-packages/web/debugerror.py", line 313, in debugerror
    return web._InternalError(djangoerror())
  File "/usr/local/lib/python3.6/site-packages/web/debugerror.py", line 303, in djangoerror
    return t(exception_type, exception_value, frames)
  File "/usr/local/lib/python3.6/site-packages/web/template.py", line 904, in __call__
    return BaseTemplate.__call__(self, *a, **kw)
  File "/usr/local/lib/python3.6/site-packages/web/template.py", line 831, in __call__
    return self.t(*a, **kw)
  File "/usr/local/lib/python3.6/site-packages/web/debugerror.py", line 196, in __template__
    </ul>
  File "/usr/local/lib/python3.6/site-packages/web/webapi.py", line 487, in cookies
    ctx._parsed_cookies = parse_cookies(http_cookie)
  File "/usr/local/lib/python3.6/site-packages/web/webapi.py", line 447, in parse_cookies
    cookie = Cookie.SimpleCookie()
NameError: name 'Cookie' is not defined
```

Did I miss another configuration step? I definitely installed python3 per the docs.

**Comment by anthonygego** [comment]
This last problem is a known issue with webpy. I don't exactly know what configurations are trigerring this and how large is the impact. The problem has already been reported in #208 and fixes were suggested at webpy/webpy#419 and webpy/webpy#421. I have no idea when this is going to be merged. We are more and more inclined to move away from webpy for its lack of maintenance, but this would be a really huge work.

**Comment by CaptainJava01** [comment]
Thanks, just found it out.
